docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,339 @@
+# ADR 033: Protobuf-based Inter-Module Communication
+
+## Changelog
+
+- 2020-10-05: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR introduces a system for permissioned inter-module communication leveraging the protobuf `Query` and `Msg`
+service definitions defined in [ADR 021](./adr-021-protobuf-query-encoding.md) and
+[ADR 031](./adr-031-msg-service.md) which provides:
+- stable module interfaces to eventually replace the keeper paradigm based on protobuf
+- stronger inter-module object capabilities guarantees
+- module accounts and sub-account authorization
+
+## Context
+
+In the current Cosmos SDK documentation on the [Object-Capability Model](../docs/core/ocap.md), it is state that:
+
+> We assume that a thriving ecosystem of Cosmos-SDK modules that are easy to compose into a blockchain application will contain faulty or malicious modules.
+
+There is currently not a thriving ecosystem of Cosmos SDK modules. We hypothesize that this is in part due to:
+1. lack of a stable v1.0 Cosmos SDK to build modules off of. Module interfaces are changing, sometimes dramatically, from
+point release to point release, often for good reasons, but this does not create a stable foundation to build on.
+2. lack of a properly implemented object capability or even object-oriented encapsulation system which makes refactors
```

Comment:
Well, I think the fact that IBC needs to prevent other modules from mess with their (`ibc/`) denoms one case: https://github.com/cosmos/cosmos-sdk/issues/7093#issuecomment-726675035. Basically because `x/bank` exposes so much, it makes a refactor constraining that interface inevitable IMHO.

Reply:
So starting from the point of view that inter-module calls should be authenticated would make that sort of refactoring unneeded. If everything were authenticated to start, we would never need to address it as a security concern later.